File: applications/api/services/customer_contract.py
"""Customer contract gate for official PDF/DOCX. Presentation/routing only."""

from __future__ import annotations

from typing import Any, Literal, Mapping

from applications.api.services.result_identity import CUSTOMER_USEFUL_GOD_CONTRACT

CustomerContractStatus = Literal["ok", "mismatch", "unversioned", "incomplete"]

EMPTY_RESULT_MESSAGE = (
    "Chưa có kết quả phân tích. Vui lòng nhập thông tin ngày giờ sinh để bắt đầu."
)
CONTRACT_MISMATCH_MESSAGE = (
    "Kết quả này được tạo bởi phiên bản dữ liệu cũ. Vui lòng phân tích lại để cập nhật kết quả."
)
CONTRACT_INCOMPLETE_MESSAGE = (
    "Kết quả phân tích chưa đủ hợp đồng hiển thị. Vui lòng phân tích lại."
)
HISTORY_MISMATCH_MESSAGE = (
    "Kết quả lịch sử không khớp với mã phân tích đang chọn. Vui lòng mở lại kết quả đã lưu."
)
RENDERER_FAILURE_MESSAGE = "Không tạo được tệp xuất. Vui lòng thử lại."


def _mapping(value: Any) -> dict[str, Any]:
    if isinstance(value, Mapping):
        return dict(value)
    return {}


def _text(value: Any) -> str:
    return str(value).strip() if value is not None else ""


def read_customer_contract(data: Mapping[str, Any] | None) -> str:
    """Return published UsefulGodView contract string, if any."""
    payload = _mapping(data)
    source = _mapping(payload.get("useful_god_source"))
    meta = _mapping(payload.get("result_meta"))
    return _text(source.get("contract")) or _text(meta.get("customer_contract"))


def customer_contract_status(data: Mapping[str, Any] | None) -> CustomerContractStatus:
    """Classify a stored analysis payload against UsefulGodView@1.5."""
    if not isinstance(data, Mapping) or not data:
        return "incomplete"
    contract = read_customer_contract(data)
    if not contract:
        if data.get("useful_god") or data.get("pattern") or data.get("bazi"):
            return "unversioned"
        return "incomplete"
    if contract != CUSTOMER_USEFUL_GOD_CONTRACT:
        return "mismatch"
    useful = _mapping(data.get("useful_god"))
    if useful.get("overall_incomplete"):
        return "ok"
    display = _text(useful.get("useful_display")) or _text(useful.get("favorable_display"))
    if not display:
        return "incomplete"
    return "ok"


def customer_contract_message(status: CustomerContractStatus) -> str:
    """Customer-safe notice for a contract status."""
    if status == "ok":
        return ""
    if status == "incomplete":
        return CONTRACT_INCOMPLETE_MESSAGE
    return CONTRACT_MISMATCH_MESSAGE


def is_compatible_customer_contract(data: Mapping[str, Any] | None) -> bool:
    """True when official PDF/DOCX may be generated."""
    return customer_contract_status(data) == "ok"

File: applications/customer_portal/src/resultState/customerContract.ts
/**
 * Frozen customer Useful God contract for Portal binding.
 * Presentation / routing only — does not recompute engines.
 */

import type { AnalysisDataDto } from "../models";

export const CUSTOMER_USEFUL_GOD_CONTRACT = "analysis_result.UsefulGodView@1.5";
export const GATE_CORE_FREEZE = "G1";
export const MONTH_PILLAR_STANDARD = "BTE-MONTH-PILLAR-LUNAR-V1.0";

export const EMPTY_RESULT_MESSAGE =
  "Chưa có kết quả phân tích. Vui lòng nhập thông tin ngày giờ sinh để bắt đầu.";

export const CONTRACT_MISMATCH_MESSAGE =
  "Kết quả này được tạo bởi phiên bản dữ liệu cũ. Vui lòng phân tích lại để cập nhật kết quả.";

export const CONTRACT_INCOMPLETE_MESSAGE =
  "Kết quả phân tích chưa đủ hợp đồng hiển thị. Vui lòng phân tích lại.";

export const MISSING_HISTORY_MESSAGE = "Không tìm thấy hồ sơ.";

export const CORRUPT_HISTORY_MESSAGE = "Không tải được kết quả đã lưu.";

export type CustomerContractStatus = "ok" | "mismatch" | "unversioned" | "incomplete";

function asRecord(value: unknown): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) return {};
  return value as Record<string, unknown>;
}

function text(value: unknown): string {
  return typeof value === "string" ? value.trim() : "";
}

/**
 * Published customer contract string, if any.
 */
export function readCustomerContract(
  data: AnalysisDataDto | null | undefined,
): string | null {
  if (!data || typeof data !== "object") return null;
  const source = asRecord(data.useful_god_source);
  const meta = asRecord(data.result_meta);
  const contract = text(source.contract) || text(meta.customer_contract);
  return contract || null;
}

/**
 * Classify stored/API payload against UsefulGodView@1.5.
 */
export function customerContractStatus(
  data: AnalysisDataDto | null | undefined,
): CustomerContractStatus {
  if (!data || typeof data !== "object") return "incomplete";
  const contract = readCustomerContract(data);
  if (!contract) {
    if (data.useful_god || data.pattern || data.calendar || data.bazi) {
      return "unversioned";
    }
    return "incomplete";
  }
  if (contract !== CUSTOMER_USEFUL_GOD_CONTRACT) return "mismatch";
  const useful = asRecord(data.useful_god);
  if (useful.overall_incomplete) return "ok";
  const hasDisplay = Boolean(text(useful.useful_display) || text(useful.favorable_display));
  if (!hasDisplay) return "incomplete";
  return "ok";
}

export function customerContractMessage(status: CustomerContractStatus): string {
  if (status === "ok") return "";
  if (status === "incomplete") return CONTRACT_INCOMPLETE_MESSAGE;
  return CONTRACT_MISMATCH_MESSAGE;
}

export function isCompatibleCustomerContract(
  data: AnalysisDataDto | null | undefined,
): boolean {
  return customerContractStatus(data) === "ok";
}
